src/marion/marion/certificates/manufactories/base.py

```diff
@@ -0,0 +1,190 @@
+import re
+import uuid
+from abc import ABC, abstractmethod
+from pathlib import Path
+from django.conf import settings
+from django.template import Context
+from django.template.engine import Engine
+from django.utils.functional import cached_property
+from jsonschema import validate, ValidationError
+from weasyprint import HTML, CSS
+from weasyprint.document import DocumentMetadata
+from weasyprint.fonts import FontConfiguration
+
+CERTIFICATES_TEMPLATE_ROOT = Path("marion/certificates")
+CERTIFICATES_ROOT = Path(settings.MEDIA_ROOT)
+
+
+class AbstractCertificate(ABC):
+
+    identifier = None
+    certificate_path = None
+
+    # Templates
+    context = None
+    css = None
+    css_template_path = None
+    html = None
+    html_template_path = None
+    template_engine = None
+
+    # Schemas
+    context_schema = None
+    init_schema = None
+
+    # Metadata
+    metadata = None
+    attachments = None
+    authors = None
+    created = None
+    description = None
+    # FIXME: add a version number
+    generator = "Marion, version xxx"
+    keywords = None
+    modified = None
+    title = None
+
+    def __init__(self, *args, **kwargs):
+
+        self.validate_init(kwargs)
+        self.identifier = kwargs.get("identifier", None) or self.get_identifier()
```

Comment:
Move all the logic to the get_identifier method so that it can be fully customized without any side effects:
```suggestion
        self.identifier = self.get_identifier(kwargs)
```